// @ltonetwork/lto/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATEMENT_TYPE_DEACTIVATE_DID = exports.ASSOCIATION_TYPE_DID_DISABLE_CAPABILITY = exports.ASSOCIATION_TYPE_DID_VERIFICATION_METHOD = exports.DEFAULT_MESSAGE_TYPE = exports.DEFAULT_DERIVATION_PATH = exports.DEFAULT_RELAY_SERVICE = exports.DEFAULT_TESTNET_NODE = exports.DEFAULT_MAINNET_NODE = exports.SEED_ENCRYPTION_ROUNDS = exports.PUBLIC_KEY_LENGTH_ECDSA = exports.UNCOMPRESSED_PUBLIC_KEY_LENGTH_ECDSA = exports.PUBLIC_KEY_LENGTH = exports.ADDRESS_VERSION = exports.EVENT_CHAIN_VERSION = void 0;
exports.EVENT_CHAIN_VERSION = 0x40;
exports.ADDRESS_VERSION = 0x1;
exports.PUBLIC_KEY_LENGTH = 32;
exports.UNCOMPRESSED_PUBLIC_KEY_LENGTH_ECDSA = 65;
exports.PUBLIC_KEY_LENGTH_ECDSA = 33;
exports.SEED_ENCRYPTION_ROUNDS = 5000;
exports.DEFAULT_MAINNET_NODE = 'https://nodes.lto.network';
exports.DEFAULT_TESTNET_NODE = 'https://testnet.lto.network';
exports.DEFAULT_RELAY_SERVICE = 'https://relay.lto.network';
exports.DEFAULT_DERIVATION_PATH = "m/44'/118'/0'/0/0";
exports.DEFAULT_MESSAGE_TYPE = 'basic';
exports.ASSOCIATION_TYPE_DID_VERIFICATION_METHOD = 0x100;
exports.ASSOCIATION_TYPE_DID_DISABLE_CAPABILITY = 0x108;
exports.STATEMENT_TYPE_DEACTIVATE_DID = 0x120;
