// @ltonetwork/lto/lib/cjs/transactions/Data.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Transaction_1 = __importDefault(require("./Transaction"));
const utils_1 = require("@noble/hashes/utils");
const base_1 = require("@scure/base");
const convert = __importStar(require("../utils/convert"));
const crypto_1 = require("../utils/crypto");
const DataEntry_1 = __importStar(require("./DataEntry"));
const BASE_FEE = 50000000;
const VAR_FEE = 10000000;
const VAR_BYTES = 256;
const DEFAULT_VERSION = 3;
class Data extends Transaction_1.default {
    constructor(data) {
        super(Data.TYPE, DEFAULT_VERSION);
        this.data = [];
        this.data = Array.isArray(data) ? data : (0, DataEntry_1.dictToData)(data);
        this.fee = BASE_FEE + Math.ceil(this.dataToBinary().length / VAR_BYTES) * VAR_FEE;
    }
    dataToBinary() {
        return this.data.reduce((binary, entry) => (0, utils_1.concatBytes)(binary, entry.toBinary()), new Uint8Array());
    }
    toBinaryV3() {
        return (0, utils_1.concatBytes)(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([(0, crypto_1.keyTypeId)(this.senderKeyType)]), base_1.base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), convert.shortToByteArray(this.data.length), this.dataToBinary());
    }
    toBinary() {
        if (!this.sender)
            throw Error('Transaction sender not set');
        switch (this.version) {
            case 3:
                return this.toBinaryV3();
            default:
                throw new Error('Incorrect version');
        }
    }
    toJSON() {
        return {
            id: this.id,
            type: this.type,
            version: this.version,
            sender: this.sender,
            senderKeyType: this.senderKeyType,
            senderPublicKey: this.senderPublicKey,
            sponsor: this.sponsor,
            sponsorKeyType: this.sponsorKeyType,
            sponsorPublicKey: this.sponsorPublicKey,
            fee: this.fee,
            timestamp: this.timestamp,
            data: this.data.map((entry) => entry.toJSON()),
            proofs: this.proofs,
            height: this.height,
        };
    }
    get dict() {
        const dictionary = {};
        this.data.forEach((entry) => (dictionary[entry.key] = entry.value));
        return dictionary;
    }
    static from(data) {
        const tx = new Data([]).initFrom(data);
        tx.data = (data.data ?? []).map(DataEntry_1.default.from);
        return tx;
    }
}
Data.TYPE = 12;
exports.default = Data;
