// @ltonetwork/lto/lib/cjs/transactions/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.txFromData = exports.Statement = exports.MappedAnchor = exports.Burn = exports.Transfer = exports.CancelSponsorship = exports.Sponsorship = exports.Register = exports.MassTransfer = exports.CancelLease = exports.Lease = exports.Data = exports.RevokeAssociation = exports.Association = exports.Anchor = exports.Transaction = void 0;
const Transaction_1 = __importDefault(require("./Transaction"));
exports.Transaction = Transaction_1.default;
const Anchor_1 = __importDefault(require("./Anchor"));
exports.Anchor = Anchor_1.default;
const Association_1 = __importDefault(require("./Association"));
exports.Association = Association_1.default;
const RevokeAssociation_1 = __importDefault(require("./RevokeAssociation"));
exports.RevokeAssociation = RevokeAssociation_1.default;
const Data_1 = __importDefault(require("./Data"));
exports.Data = Data_1.default;
const Lease_1 = __importDefault(require("./Lease"));
exports.Lease = Lease_1.default;
const CancelLease_1 = __importDefault(require("./CancelLease"));
exports.CancelLease = CancelLease_1.default;
const MassTransfer_1 = __importDefault(require("./MassTransfer"));
exports.MassTransfer = MassTransfer_1.default;
const Register_1 = __importDefault(require("./Register"));
exports.Register = Register_1.default;
const Sponsorship_1 = __importDefault(require("./Sponsorship"));
exports.Sponsorship = Sponsorship_1.default;
const CancelSponsorship_1 = __importDefault(require("./CancelSponsorship"));
exports.CancelSponsorship = CancelSponsorship_1.default;
const Transfer_1 = __importDefault(require("./Transfer"));
exports.Transfer = Transfer_1.default;
const Burn_1 = __importDefault(require("./Burn"));
exports.Burn = Burn_1.default;
const MappedAnchor_1 = __importDefault(require("./MappedAnchor"));
exports.MappedAnchor = MappedAnchor_1.default;
const Statement_1 = __importDefault(require("./Statement"));
exports.Statement = Statement_1.default;
function txFromData(data) {
    switch (data.type) {
        case 4:
            return Transfer_1.default.from(data);
        case 8:
            return Lease_1.default.from(data);
        case 9:
            return CancelLease_1.default.from(data);
        case 11:
            return MassTransfer_1.default.from(data);
        case 12:
            return Data_1.default.from(data);
        case 15:
            return Anchor_1.default.from(data);
        case 16:
            return Association_1.default.from(data);
        case 17:
            return RevokeAssociation_1.default.from(data);
        case 18:
            return Sponsorship_1.default.from(data);
        case 19:
            return CancelSponsorship_1.default.from(data);
        case 20:
            return Register_1.default.from(data);
        case 21:
            return Burn_1.default.from(data);
        case 22:
            return MappedAnchor_1.default.from(data);
        case 23:
            return Statement_1.default.from(data);
        default:
            throw Error('Transaction type not recognized');
    }
}
exports.txFromData = txFromData;
