// @ltonetwork/lto/lib/cjs/transactions/RevokeAssociation.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const Transaction_1 = __importDefault(require("./Transaction"));
const utils_1 = require("@noble/hashes/utils");
const base_1 = require("@scure/base");
const convert = __importStar(require("../utils/convert"));
const crypto_1 = require("../utils/crypto");
const Binary_1 = __importDefault(require("../Binary"));
const DEFAULT_FEE = 50000000;
const DEFAULT_VERSION = 4;
class RevokeAssociation extends Transaction_1.default {
    constructor(associationType, recipient, subject) {
        super(RevokeAssociation.TYPE, DEFAULT_VERSION, DEFAULT_FEE);
        this.recipient = typeof recipient === 'string' ? recipient : recipient.address;
        this.associationType = associationType;
        if (subject)
            this.subject = new Binary_1.default(subject);
    }
    toBinaryV1() {
        const hashBytes = this.subject
            ? (0, utils_1.concatBytes)(Uint8Array.from([1]), convert.shortToByteArray(this.subject.length), Uint8Array.from(this.subject))
            : Uint8Array.from([0]);
        return (0, utils_1.concatBytes)(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), base_1.base58.decode(this.senderPublicKey), base_1.base58.decode(this.recipient), convert.integerToByteArray(this.associationType), hashBytes, convert.longToByteArray(this.timestamp), convert.longToByteArray(this.fee));
    }
    toBinaryV3() {
        return (0, utils_1.concatBytes)(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([(0, crypto_1.keyTypeId)(this.senderKeyType)]), base_1.base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), base_1.base58.decode(this.recipient), convert.integerToByteArray(this.associationType), convert.shortToByteArray(this.subject?.length ?? 0), this.subject ?? new Uint8Array());
    }
    toBinaryV4() {
        return (0, utils_1.concatBytes)(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([(0, crypto_1.keyTypeId)(this.senderKeyType)]), base_1.base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), convert.longToByteArray(this.associationType), base_1.base58.decode(this.recipient), convert.shortToByteArray(this.subject?.length ?? 0), this.subject ?? new Uint8Array());
    }
    toBinary() {
        if (!this.sender)
            throw Error('Transaction sender not set');
        switch (this.version) {
            case 1:
                return this.toBinaryV1();
            case 3:
                return this.toBinaryV3();
            case 4:
                return this.toBinaryV4();
            default:
                throw new Error('Incorrect version');
        }
    }
    toJSON() {
        return {
            id: this.id,
            type: this.type,
            version: this.version,
            sender: this.sender,
            senderKeyType: this.senderKeyType,
            senderPublicKey: this.senderPublicKey,
            sponsor: this.sponsor,
            sponsorKeyType: this.sponsorKeyType,
            sponsorPublicKey: this.sponsorPublicKey,
            recipient: this.recipient,
            associationType: this.associationType,
            fee: this.fee,
            timestamp: this.timestamp,
            subject: this.subject?.base58,
            proofs: this.proofs,
            height: this.height,
        };
    }
    static from(data) {
        return new RevokeAssociation(data.associationType, data.recipient, data.subject ? Binary_1.default.fromBase58(data.subject) : undefined).initFrom(data);
    }
}
RevokeAssociation.TYPE = 17;
exports.default = RevokeAssociation;
