// @ltonetwork/lto/lib/esm/accounts/ed25519/ED25519.js
import { Cypher } from '../Cypher.js';
import nacl from 'tweetnacl';
import { blake2b } from '@noble/hashes/blake2b';
import { DecryptError } from '../../errors/index.js';
export class ED25519 extends Cypher {
    constructor(sign, encrypt) {
        super('ed25519');
        this.sign = sign;
        this.encrypt = encrypt;
    }
    static sealNonce(epk, publicKey) {
        return blake2b.create({ dkLen: nacl.box.nonceLength }).update(epk).update(publicKey).digest();
    }
    static seal(message, publicKey) {
        const ekp = nacl.box.keyPair();
        const out = new Uint8Array(message.length + nacl.box.overheadLength + nacl.box.publicKeyLength);
        out.set(ekp.publicKey, 0);
        const nonce = this.sealNonce(ekp.publicKey, publicKey);
        const ct = nacl.box(message, nonce, publicKey, ekp.secretKey);
        out.set(ct, nacl.box.publicKeyLength);
        return out;
    }
    static sealOpen(ciphertext, publicKey, secretKey) {
        const epk = ciphertext.slice(0, nacl.box.publicKeyLength);
        ciphertext = ciphertext.slice(nacl.box.publicKeyLength);
        const nonce = this.sealNonce(epk, publicKey);
        return nacl.box.open(ciphertext, nonce, epk, secretKey);
    }
    encryptMessage(input) {
        return ED25519.seal(input, this.encrypt.publicKey);
    }
    decryptMessage(input) {
        if (!this.encrypt.privateKey)
            throw new Error('Missing private key for decryption');
        const output = ED25519.sealOpen(input, this.encrypt.publicKey, this.encrypt.privateKey);
        if (!output)
            throw new DecryptError('Unable to decrypt message with given keys');
        return output;
    }
    createSignature(input) {
        if (!this.sign.privateKey)
            throw new Error('Missing private key for signing');
        return nacl.sign.detached(input, this.sign.privateKey);
    }
    verifySignature(input, signature) {
        return signature.length === 64 && nacl.sign.detached.verify(input, signature, this.sign.publicKey);
    }
}
