// @ltonetwork/lto/lib/esm/transactions/CancelSponsorship.js
import Transaction from './Transaction.js';
import { concatBytes } from '@noble/hashes/utils';
import { base58 } from '@scure/base';
import * as convert from '../utils/convert.js';
import { keyTypeId } from '../utils/crypto.js';
const DEFAULT_FEE = 500000000;
const DEFAULT_VERSION = 3;
class CancelSponsorship extends Transaction {
    constructor(recipient) {
        super(CancelSponsorship.TYPE, DEFAULT_VERSION, DEFAULT_FEE);
        this.recipient = typeof recipient === 'string' ? recipient : recipient.address;
    }
    toBinaryV1() {
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), base58.decode(this.senderPublicKey), base58.decode(this.recipient), convert.longToByteArray(this.timestamp), convert.longToByteArray(this.fee));
    }
    toBinaryV3() {
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([keyTypeId(this.senderKeyType)]), base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), base58.decode(this.recipient));
    }
    toBinary() {
        if (!this.sender)
            throw Error('Transaction sender not set');
        switch (this.version) {
            case 1:
                return this.toBinaryV1();
            case 3:
                return this.toBinaryV3();
            default:
                throw Error('Incorrect version');
        }
    }
    toJSON() {
        return {
            id: this.id,
            type: this.type,
            version: this.version,
            sender: this.sender,
            senderKeyType: this.senderKeyType,
            senderPublicKey: this.senderPublicKey,
            sponsor: this.sponsor,
            sponsorKeyType: this.sponsorKeyType,
            sponsorPublicKey: this.sponsorPublicKey,
            recipient: this.recipient,
            timestamp: this.timestamp,
            fee: this.fee,
            proofs: this.proofs,
            height: this.height,
        };
    }
    static from(data) {
        return new CancelSponsorship(data.recipient).initFrom(data);
    }
}
CancelSponsorship.TYPE = 19;
export default CancelSponsorship;
