// @ltonetwork/lto/lib/esm/transactions/RevokeAssociation.js
import Transaction from './Transaction.js';
import { concatBytes } from '@noble/hashes/utils';
import { base58 } from '@scure/base';
import * as convert from '../utils/convert.js';
import { keyTypeId } from '../utils/crypto.js';
import Binary from '../Binary.js';
const DEFAULT_FEE = 50000000;
const DEFAULT_VERSION = 4;
class RevokeAssociation extends Transaction {
    constructor(associationType, recipient, subject) {
        super(RevokeAssociation.TYPE, DEFAULT_VERSION, DEFAULT_FEE);
        this.recipient = typeof recipient === 'string' ? recipient : recipient.address;
        this.associationType = associationType;
        if (subject)
            this.subject = new Binary(subject);
    }
    toBinaryV1() {
        const hashBytes = this.subject
            ? concatBytes(Uint8Array.from([1]), convert.shortToByteArray(this.subject.length), Uint8Array.from(this.subject))
            : Uint8Array.from([0]);
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), base58.decode(this.senderPublicKey), base58.decode(this.recipient), convert.integerToByteArray(this.associationType), hashBytes, convert.longToByteArray(this.timestamp), convert.longToByteArray(this.fee));
    }
    toBinaryV3() {
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([keyTypeId(this.senderKeyType)]), base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), base58.decode(this.recipient), convert.integerToByteArray(this.associationType), convert.shortToByteArray(this.subject?.length ?? 0), this.subject ?? new Uint8Array());
    }
    toBinaryV4() {
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([keyTypeId(this.senderKeyType)]), base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), convert.longToByteArray(this.associationType), base58.decode(this.recipient), convert.shortToByteArray(this.subject?.length ?? 0), this.subject ?? new Uint8Array());
    }
    toBinary() {
        if (!this.sender)
            throw Error('Transaction sender not set');
        switch (this.version) {
            case 1:
                return this.toBinaryV1();
            case 3:
                return this.toBinaryV3();
            case 4:
                return this.toBinaryV4();
            default:
                throw new Error('Incorrect version');
        }
    }
    toJSON() {
        return {
            id: this.id,
            type: this.type,
            version: this.version,
            sender: this.sender,
            senderKeyType: this.senderKeyType,
            senderPublicKey: this.senderPublicKey,
            sponsor: this.sponsor,
            sponsorKeyType: this.sponsorKeyType,
            sponsorPublicKey: this.sponsorPublicKey,
            recipient: this.recipient,
            associationType: this.associationType,
            fee: this.fee,
            timestamp: this.timestamp,
            subject: this.subject?.base58,
            proofs: this.proofs,
            height: this.height,
        };
    }
    static from(data) {
        return new RevokeAssociation(data.associationType, data.recipient, data.subject ? Binary.fromBase58(data.subject) : undefined).initFrom(data);
    }
}
RevokeAssociation.TYPE = 17;
export default RevokeAssociation;
