// @ltonetwork/lto/lib/esm/transactions/Transfer.js
import Transaction from './Transaction.js';
import { concatBytes } from '@noble/hashes/utils';
import { base58 } from '@scure/base';
import * as convert from '../utils/convert.js';
import { keyTypeId } from '../utils/crypto.js';
import Binary from '../Binary.js';
const DEFAULT_FEE = 100000000;
const DEFAULT_VERSION = 3;
class Transfer extends Transaction {
    constructor(recipient, amount, attachment = '') {
        super(Transfer.TYPE, DEFAULT_VERSION, DEFAULT_FEE);
        this.recipient = typeof recipient === 'string' ? recipient : recipient.address;
        this.amount = amount;
        this.attachment = new Binary(attachment);
    }
    toBinaryV2() {
        return concatBytes(Uint8Array.from([this.type, this.version]), base58.decode(this.senderPublicKey), convert.longToByteArray(this.timestamp), convert.longToByteArray(this.amount), convert.longToByteArray(this.fee), base58.decode(this.recipient), convert.shortToByteArray(this.attachment.length), this.attachment);
    }
    toBinaryV3() {
        return concatBytes(Uint8Array.from([this.type, this.version]), convert.stringToByteArray(this.networkId), convert.longToByteArray(this.timestamp), Uint8Array.from([keyTypeId(this.senderKeyType)]), base58.decode(this.senderPublicKey), convert.longToByteArray(this.fee), base58.decode(this.recipient), convert.longToByteArray(this.amount), convert.shortToByteArray(this.attachment.length), this.attachment);
    }
    toBinary() {
        if (!this.sender)
            throw Error('Transaction sender not set');
        switch (this.version) {
            case 2:
                return this.toBinaryV2();
            case 3:
                return this.toBinaryV3();
            default:
                throw Error('Incorrect version');
        }
    }
    toJSON() {
        return {
            id: this.id,
            type: this.type,
            version: this.version,
            sender: this.sender,
            senderKeyType: this.senderKeyType,
            senderPublicKey: this.senderPublicKey,
            sponsor: this.sponsor,
            sponsorKeyType: this.sponsorKeyType,
            sponsorPublicKey: this.sponsorPublicKey,
            fee: this.fee,
            timestamp: this.timestamp,
            amount: this.amount,
            recipient: this.recipient,
            attachment: this.attachment.base58,
            proofs: this.proofs,
            height: this.height,
        };
    }
    static from(data) {
        const attachment = data.attachment ? Binary.fromBase58(data.attachment) : '';
        return new Transfer(data.recipient, data.amount, attachment).initFrom(data);
    }
}
Transfer.TYPE = 4;
export default Transfer;
